lib/src/subscription.dart

```diff
@@ -0,0 +1,155 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+import 'dart:collection';
+
+import 'native/realm_core.dart';
+import 'realm_class.dart';
+import 'util.dart';
+
+class Subscription {
+  final SubscriptionHandle _handle;
+
+  Subscription._(this._handle);
+
+  @override
+  // ignore: hash_and_equals
+  bool operator ==(Object other) {
+    if (identical(this, other)) return true;
+    if (other is! Subscription) return false;
+    return realmCore.subscriptionEquals(this, other);
+  }
+}
+
+extension SubscriptionInternal on Subscription {
+  SubscriptionHandle get handle => _handle;
+}
+
+class _SubscriptionIterator implements Iterator<Subscription> {
+  int _index = -1;
+  final SubscriptionSet _subscriptions;
+
+  _SubscriptionIterator._(this._subscriptions);
+
+  @override
+  Subscription get current => _subscriptions.elementAt(_index);
+
+  @override
+  bool moveNext() => ++_index < _subscriptions.length;
+}
+
+enum SubscriptionSetState {
+  uncommitted,
+  pending,
+  bootstrapping,
+  complete,
+  error,
+  superseded,
+}
+
+abstract class SubscriptionSet with IterableMixin<Subscription> {
+  Realm _realm;
+  SubscriptionSetHandle _handle;
+
+  SubscriptionSet._(this._realm, this._handle);
+
+  Subscription? find<T extends RealmObject>(RealmResults<T> query) {
+    return realmCore.findSubscriptionByQuery(this, query).convert(Subscription._);
+  }
+
+  Subscription? findByName(String name) {
+    return realmCore.findSubscriptionByName(this, name).convert(Subscription._);
+  }
+
+  Future<SubscriptionSetState> waitForStateChange(SubscriptionSetState state) async {
+    return SubscriptionSetState.values[await realmCore.waitForSubscriptionSetStateChange(this, state)];
+  }
+
+  @override
+  int get length => realmCore.getSubscriptionSetSize(this);
+
+  @override
+  Subscription elementAt(int index) {
+    return Subscription._(realmCore.subscriptionAt(this, index));
+  }
+
+  Subscription operator [](int index) => elementAt(index);
+
+  @override
+  _SubscriptionIterator get iterator => _SubscriptionIterator._(this);
+
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action);
+
+  int get version => realmCore.subscriptionSetVersion(this);
+
+  SubscriptionSetState get state => SubscriptionSetState.values[realmCore.subscriptionSetState(this)];
+}
+
+extension SubscriptionSetInternal on SubscriptionSet {
+  Realm get realm => _realm;
+  SubscriptionSetHandle get handle => _handle;
+
+  static SubscriptionSet create(Realm realm, SubscriptionSetHandle handle) => MutableSubscriptionSet._(realm, handle);
```

Comment:
What do you find hard to read?

Reply:
I find it confusing that we appear to have two types of sets, when in reality we always use the mutable one and just pretend that it's immutable. If a user inspects the value at runtime, they're likely to see that they're operating on a mutable set, even outside the update block. If we want to go for type safety, I'd suggest we use actual different instances rather than the same instance.

Reply:
Okay, I will make `SubscriptionSet` abstract. Let the factory return a `_ImmutableSubscriptionSet` that overloads `update` but creates an actual new `MutableSubscriptionSet`